File: 3203_find_minimum_diameter_after_merging_two_trees.py
class Solution(object):
    def calculate_diameter(self, edges):
        from collections import defaultdict, deque
        graph = defaultdict(list)
        degree = [0] * (len(edges) + 1)
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
            degree[u] += 1
            degree[v] += 1
        queue = deque(i for i in range(len(degree)) if degree[i] == 1)
        remaining = len(edges) + 1
        depth = 0
        while remaining > 2:
            size = len(queue)
            remaining -= size
            depth += 1
            for _ in range(size):
                node = queue.popleft()
                for neighbor in graph[node]:
                    degree[neighbor] -= 1
                    if degree[neighbor] == 1:
                        queue.append(neighbor)
        return 2 * depth + 1 if remaining == 2 else 2 * depth

    def minimumDiameterAfterMerge(self, edges1, edges2):
        d1 = self.calculate_diameter(edges1)
        d2 = self.calculate_diameter(edges2)
        return max(d1, d2, (d1 + 1) // 2 + (d2 + 1) // 2 + 1)

File: 3203_find_minimum_diameter_after_merging_two_trees.java
class Solution {
    private int calculateDiameter(int[][] edges) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] degree = new int[edges.length + 1];
        for (int[] edge : edges) {
            graph.computeIfAbsent(edge[0], k -> new ArrayList<>()).add(edge[1]);
            graph.computeIfAbsent(edge[1], k -> new ArrayList<>()).add(edge[0]);
            degree[edge[0]]++;
            degree[edge[1]]++;
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < degree.length; i++) {
            if (degree[i] == 1) queue.add(i);
        }
        int nodes = edges.length + 1, depth = 0;
        while (nodes > 2) {
            int size = queue.size();
            nodes -= size;
            depth++;
            while (size-- > 0) {
                int node = queue.poll();
                for (int neighbor : graph.get(node)) {
                    if (--degree[neighbor] == 1) queue.add(neighbor);
                }
            }
        }
        return nodes == 2 ? 2 * depth + 1 : 2 * depth;
    }

    public int minimumDiameterAfterMerge(int[][] edges1, int[][] edges2) {
        int d1 = calculateDiameter(edges1);
        int d2 = calculateDiameter(edges2);
        return Math.max(d1, Math.max(d2, (d1 + 1) / 2 + (d2 + 1) / 2 + 1));
    }
}
